Assets/MixedRealityToolkit.Tests/PlayModeTests/SpeechTests.cs

```diff
@@ -0,0 +1,77 @@
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+#if !WINDOWS_UWP
+// When the .NET scripting backend is enabled and C# projects are built
+// The assembly that this file is part of is still built for the player,
+// even though the assembly itself is marked as a test assembly (this is not
+// expected because test assemblies should not be included in player builds).
+// Because the .NET backend is deprecated in 2018 and removed in 2019 and this
+// issue will likely persist for 2018, this issue is worked around by wrapping all
+// play mode tests in this check.
+
+using Microsoft.MixedReality.Toolkit.Diagnostics;
+using Microsoft.MixedReality.Toolkit.Input;
+using Microsoft.MixedReality.Toolkit.Utilities;
+using NUnit.Framework;
+using System;
+using System.Collections;
+using System.Linq;
+using UnityEngine;
+using UnityEngine.TestTools;
+
+namespace Microsoft.MixedReality.Toolkit.Tests
+{
+    public class SpeechTests
+    {
+        [SetUp]
+        public void Setup()
+        {
+            PlayModeTestUtilities.Setup();
+        }
+
+        [TearDown]
+        public void TearDown()
+        {
+            PlayModeTestUtilities.TearDown();
+        }
+
+        [UnityTest]
+        public IEnumerator TestToggleProfilerCommand()
+        {
+            int frameDelay = 10;
+
+            // Confirm that the diagnostics system is enabled.
+            IMixedRealityDiagnosticsSystem diagnosticsSystem = null;
+            MixedRealityServiceRegistry.TryGetService<IMixedRealityDiagnosticsSystem>(out diagnosticsSystem);
+            Assert.IsNotNull(diagnosticsSystem, "The diagnostics system is not enabled in the scene.");
+            yield return null;
+
+            // This test uses the input system to simulate speech commands.
+            IMixedRealityInputSystem inputSystem = null;
+            MixedRealityServiceRegistry.TryGetService<IMixedRealityInputSystem>(out inputSystem);
+            Assert.IsNotNull(inputSystem, "The input system is not enabled in the scene.");
+            yield return null;
+
+            // Verfiy that the VisualProfiler is enabled.
+            Assert.IsTrue(diagnosticsSystem.ShowProfiler, "The VisualProfiler is not active.");
+            yield return null;
+
+            // Toggle the profiler visualization off.
+            var gazeInputSource = inputSystem.DetectedInputSources.Where(x => x.SourceName.Equals("Gaze")).First();
+            inputSystem.RaiseSpeechCommandRecognized(
+                gazeInputSource, 
+                RecognitionConfidenceLevel.High, 
+                new TimeSpan(), 
+                DateTime.Now, 
+                new SpeechCommands("toggle profiler", KeyCode.Alpha9, MixedRealityInputAction.None));
+            // It may take a few frames before the event is handled and the system responds to the state change.
+            for (int i = 0; i < frameDelay; i++) { yield return null; }
```

Comment:
Does this fail every single time when there's only a single yield return null;?

I suppose in general there's not really any gaurantee that everything happens in a frame, so maybe having some general awaiting assert would be more helpful (i.e. it waits for the condition to be true and waits for up to one second).

The second part of the comment

Reply:
in an earlier incarnation, it failed each time unless under the debugger (timing changes)... i felt this was a more robust way ensure a good buffer of time :)